cutlery/_wordpiece.pyx

```diff
@@ -0,0 +1,84 @@
+from typing import List, Tuple
+from cython.operator cimport dereference as deref
+from libcpp cimport pair
+
+cdef struct PieceMatch:
+    bint found
+    int prefix_len
+    int piece_id
+
+cdef class WordPieceProcessor:
+    def __init__(self, pieces: List[str]):
+        for idx, piece in enumerate(pieces):
+            if piece.startswith("##"):
+                self.continuation_pieces[piece[2:].encode('utf8')] = idx
+            else:
+                self.initial_pieces[piece.encode('utf8')] = idx
+
+    def encode(self, token: str) -> Tuple[List[int], List[str]]:
+        """
+        Encode a token using word pieces. The piece identifiers are
+        returned. If no piece could be found for a suffix, the special
+        piece identifier `-1` is used.
+
+            token: The token to encode as pieces.
+        """
+        token_ids = []
+        token_pieces = []
+        cdef const unordered_map[string, size_t]* pieces = &self.initial_pieces
+        prefix = ""
+        cdef PieceMatch match
+        while token:
+            match = _find_longest_prefix(pieces, token)
+            if not match.found:
+                token_ids.append(-1)
+                token_pieces.append(None)
+                return token_ids, token_pieces
+            token_ids.append(match.piece_id)
+            token_pieces.append(f"{prefix}{token[:match.prefix_len]}")
+            token = token[match.prefix_len:]
+            pieces = &self.continuation_pieces
+            prefix = "##"
+        return token_ids, token_pieces
+
+    def get_initial(self, piece: str) -> int:
+        cdef unordered_map[string, size_t].const_iterator iter
+        iter = self.initial_pieces.const_find(piece.encode('utf8'))
+        if iter == self.initial_pieces.end():
+            raise KeyError()
```

Comment:
We could use a more descriptive error message here.

Reply:
Fixed.